iavl/iavl_tree.go

```diff
@@ -41,139 +135,303 @@ func NewIAVLTree(cacheSize int, db dbm.DB) *IAVLTree {
 		// It sets all the hashes recursively,
 		// clears all the leftNode/rightNode values recursively,
 		// and all the .persisted flags get set.
-		PanicSanity("It is unsafe to Copy() an unpersisted tree.")
-	} else if t.ndb == nil && t.root.hash == nil {
+		cmn.PanicSanity("It is unsafe to Copy() an unpersisted tree.")
+
+	} else if t.ndb == nil && root.hash == nil {
 		// An in-memory IAVLTree is finalized when the hashes are
 		// calculated.
-		t.root.hashWithCount(t)
+		root.hashWithCount(t)
 	}
+
+	lastId++
+	tmp := list.New()
+	tmp.PushBackList(t.roots)
+
 	return &IAVLTree{
-		root: t.root,
-		ndb:  t.ndb,
+		ndb:     t.ndb,
+		version: t.version,
+		roots:   tmp,
+		id:      lastId,
 	}
 }
 
+// Size returns the number of internal and leaf nodes in the tree.
 func (t *IAVLTree) Size() int {
-	if t.root == nil {
+	//fmt.Printf("Version ")
+	root := t.GetRoot(t.version)
+	if root == nil {
 		return 0
 	}
-	return t.root.size
+	return root.size
 }
 
+// Height returns the maximum tree height.
 func (t *IAVLTree) Height() int8 {
-	if t.root == nil {
+	//fmt.Printf("Version ")
+	root := t.GetRoot(t.version)
+	if root == nil {
 		return 0
 	}
-	return t.root.height
+	return root.height
+}
+
+// Version is the latest uncommitted version number.
+func (t *IAVLTree) Version() int {
+	//fmt.Printf("Version ")
+	root := t.GetRoot(t.version)
+	if root == nil {
+		return 0
+	}
+	return root.version
 }
 
 func (t *IAVLTree) Has(key []byte) bool {
-	if t.root == nil {
+	//fmt.Printf("Has ")
+	root := t.GetRoot(t.version)
+	if root == nil {
 		return false
 	}
-	return t.root.has(t, key)
+	return root.has(t, key)
 }
 
+// Proof of the latest key
 func (t *IAVLTree) Proof(key []byte) (value []byte, proofBytes []byte, exists bool) {
-	value, proof := t.ConstructProof(key)
+	//fmt.Printf("Proof ")
+	value, _, proof := t.ConstructProof(key, t.version)
 	if proof == nil {
+		//fmt.Printf("Missing Proof\n")
 		return nil, nil, false
 	}
 	proofBytes = wire.BinaryBytes(proof)
 	return value, proofBytes, true
 }
 
+// Proof of a key at a specific version
+func (t *IAVLTree) ProofVersion(key []byte, version int) (value []byte, proofBytes []byte, exists bool) {
+	//fmt.Printf("ProofVersion\n")
+	value, _, proof := t.ConstructProof(key, version)
+	if proof == nil {
+		return nil, nil, false
+	}
+	proofBytes = wire.BinaryBytes(proof)
+	return value, proofBytes, true
+}
+
+// Set a value with a new key or an existing one
 func (t *IAVLTree) Set(key []byte, value []byte) (updated bool) {
-	if t.root == nil {
-		t.root = NewIAVLNode(key, value)
+	//fmt.Printf("Set %X/%X ", key, value)
+
+	root := t.GetRoot(t.version)
+	if root == nil {
+		//fmt.Printf("Initializing Tree\n")
+
+		SetValue(t.roots, 0, NewIAVLNode(key, value, t.version))
+
 		return false
 	}
-	t.root, updated = t.root.set(t, key, value)
+
+	//fmt.Printf("Added new Root\n")
+	root.validate(t)
+	root, updated = root.set(t, key, value)
+	root.validate(t)
+
+	SetValue(t.roots, 0, root)
```

Comment:
Each time we add (or remove) a key/value to an existing tree, because the original tree is immutable (mostly :-) we end up with two perfectly formed trees. We have to add the leaf node, but also a new root and all of the internal nodes down to that leaf. Mostly, all of these new nodes are garbage collected, only a few will be persisted. 

For a Save, we are actually just preserving the last tree and then updating the database.